lib/types: add signed/unsigned integer types

Profpatsch opened the pull request:
Add types for signed/unsigned integers, as well as a `port` type.
Documentation for various types and the module in general is improved as well.

- [x] tested manually for example values
- [x] checked the rendering of the manual

Followup to #27239.

Profpatsch committed:
lib/types: add intBetween

An int type that checks the value range.

Profpatsch committed:
lib/modules: Change type error to be gramatically nicer

Before:
  <x> is not a integer between 0 and 100 (inclusively).
(notice that “a” is wrong, it should be “an”)
Now:
  <x> is not of type `integer between 0 and 100 (inclusively)'.

This sounds a bit more formal, but circumvents the grammatical problems.
Multi-word type descriptions are also easier to see.

Profpatsch committed:
lib/types: add various signed/unsigned int types

It is sometimes necessary to restrict the domain of integers for configurations,
as well as restricting them to unsigned/positive values.

Profpatsch committed:
lib/types: add port type

Profpatsch committed:
lib/types: nixos manual documentation for signed/unsinged int

Synchronize the manual for the new types.

volth commented on lib/types.nix:
```diff
@@ -107,22 +107,83 @@ rec {
+      };
 
+    /* A network port number. */
+    port = ints.unsigned16;
```
maybe ```port = ints.between 1 65535``` (excluding 0) ?

Infinisil replied:
I wouldn't consider using `ints.between 1 65535` instead of `unsigned16` an overcomplication. I'm with @volth here, it makes more sense to not include 0

grahamc replied:
I like it, and can imagine more confused users for why port doesn't accept a valid value, than confused users about why their service on port 0 doesn't work.

volth commented on lib/types.nix:
```diff
@@ -107,22 +107,83 @@ rec {
+        signed = int;
+        signed8 = sign 8 256;
+        signed16 = sign 16 65536;
+        signed32 = sign 32 4294967296;
```
I am for ```int```, ```int8```, ```int16```, ```int32```.
```signed``` is needless synonym

Profpatsch replied:
I deliberately decided against int/uint, because

* `ints.int8` is redundant
* `ints.signed8` is more specific
* more importantly, it would lead to `uint`, which is just a bad abbreviation

volth replied:
```unsigned``` or ```unsigned-int``` is not bad, but ```signed``` seems excessive when there is ```int``` already (well, there is ```signed``` in C, but nobody uses it or remember that there is such alias for ```int```)

Profpatsch replied:
So un-`signed` is intuitive while nobody knows what `signed` means?

volth replied:
Kind of. ```uint``` might be a bad abbreviation, but of those who type ```uint``` quite a few recall what it abbreviates ```u[n-signed ]int```. Just a ```U```-word no matter what is after U. ```S```-word is less known.

Profpatsch replied:
I don’t buy your reasoning. If one didn’t know what signed means, one shouldn’t know what unsigned means as well. If one knew what unsigned means, the step to signed should be trivial. `uint` is the first abbreviation of all. `u8` and `i8` are maybe easy for Rust programmers, but horrible abbreviations out of low-level system programming context.

volth replied:
https://github.com/search?l=C&q=unsigned&ref=simplesearch&type=Code&utf8=%E2%9C%93
https://github.com/search?l=C&q=signed&ref=simplesearch&type=Code&utf8=%E2%9C%93
For C, ```int``` and ```unsigned``` give almost the same number of results, ```signed``` is 10 times less often .

Profpatsch replied:
I won’t discuss this any further, since it’s too bike-sheddy for my taste.

Profpatsch replied:
I won’t discuss this any further, since it’s too bike-sheddy for my taste.

volth replied:
ok, but not only.
My first (and still) point is not about which work to prefer, but against using two words for the same entity: against ```signed = int;```. All the rest are suggesting ways on how to avoid it is opinionated and bike-sheddy.

Profpatsch commented on lib/types.nix:
```diff
@@ -107,22 +107,83 @@ rec {
+      };
 
+    /* A network port number. */
+    port = ints.unsigned16;
```
No, 0 is a valid port that means: “give me an ephemeral (some high) port”

volth replied:
In the context of service options ```0``` means ```a magic value```, for example [Value of 0 disables the keep-alive style monitoring](https://github.com/NixOS/nixpkgs/blob/56904d7c423f2b13b37fbd29f39bbb4b52bc7824/nixos/modules/services/networking/autossh.nix#L38)

that would be better ```types.nullOr types.port``` or some kind of ADT.
Non-magical port is 1..65535

Profpatsch replied:
https://www.grc.com/port_0.htm

```
"Port Zero" does not officially exist. […]
The designers of the original Berkeley UNIX "Sockets" interface, upon which much of the technology and practice we use today is based, set aside the specification of "port 0" to be used as a sort of "wild card" port. When programming the Sockets interface, the provision of a zero value is generally taken to mean "let the system choose one for me". Programmers who specify "port 0" know that it is an invalid port. They are asking the operating system to pick and assign whatever non-zero port is available and appropriate for their purpose.
```

In favour of simplicity I’d not overcomplicate the option type.
cc @aszlig

grahamc commented:
@Profpatsch can you rebase? This PR seems to be based on a fairly old master.

grahamc commented on lib/types.nix:
```diff
@@ -107,22 +107,83 @@ rec {
+            name = "intBetween";
+            description = "Integer between ${betweenDesc lowest highest}.";
+          };
+        ign = lowest: highest: name: docStart:
```
what does ign mean?

Profpatsch replied:
Helper w/ cute name only used internally in `sign` and `unsign` directly below.

grahamc commented on lib/types.nix:
```diff
@@ -107,17 +107,64 @@ rec {
+        signed16 = sign 16 65536;
+        signed32 = sign 32 4294967296;
 
-    intBetween = min: max:
```
are you intentionally removing this from the API? Perhaps a deprecation notice and map it back to int.between?

grahamc replied:
Oh, it seems you put it back.

grahamc commented on nixos/doc/manual/development/option-types.xml:
```diff
@@ -23,30 +23,74 @@
+        respectively (e.g. <literal>−128</literal> to <literal>127</literal>
+        for 8 bits).
+    </para></listitem>
+  </varlistentry>
```
Would you mind breaking these out, and specifying exactly their min/max?

Profpatsch replied:
Not sure if that’d improve the description, since those are based on byte-lengths anyway.

grahamc commented on nixos/doc/manual/development/option-types.xml:
```diff
@@ -23,30 +23,74 @@
-    <listitem><para>An integer between <replaceable>min</replaceable>
-        and <replaceable>max</replaceable> (both inclusive).
-        Useful for e.g. port ranges.
+    <listitem><para>An unsigned integer (that is ≥ 0).
```
Hmm... I'm not sure we should put unicode here, given these docs are sometimes read on TTY8.

Profpatsch replied:
Doesn’t seem to be a problem for these:
![screenshot](https://user-images.githubusercontent.com/3153638/32191363-1bb660ac-bdb1-11e7-9ca6-4b5a8e682388.png)

Do you refer to the installation-manual tty? Is that not in UTF-8?

Profpatsch replied:
Removed it, since it’s mandatory that it shows up everywhere.

Profpatsch committed:
lib/types: remove port type again

Will be introduced as a taggedUnion, once that type is in nixpkgs.

Profpatsch committed:
lib/types: signed -> s, unsigned -> u, remove signed alias

Mirrors the way it’s done in modern low-level languages like Rust (by input of
@nbp).
Removes the signed alias for int.

Profpatsch committed:
lib/types: add tests for `ints.between` and `ints.unsigned`

The int types are trivial invocations of `ints.between`, so they are not tested
explicitely.

Profpatsch committed:
lib/types: update docs to match the new type names

Profpatsch committed:
lib/types: match lowercase descriptions of other types

Profpatsch committed:
lib/types: add `ints.positive`.

For values that are positive, but cannot be 0.

nbp commented on lib/modules.nix:
```diff
@@ -338,7 +338,7 @@ rec {
     mergedValue = foldl' (res: def:
       if type.check def.value then res
-      else throw "The option value `${showOption loc}' in `${def.file}' is not a ${type.description}.")
+      else throw "The option value `${showOption loc}' in `${def.file}' is not of type `${type.description}'.")
```
From what I understand we are moving away from asymmetrical quoting?
This set of commit changes the quoting within comments as well, please revert to do only one type of quotation mark.

Profpatsch replied:
We are? My impression was Eelco would like to keep the “special” quoting?

Profpatsch replied:
```
12:03:46        niksnut | using `...' is always wrong
```

heh. :)

zimbatm replied:
Unicode quotes have practical implications that make them undesirable.

See https://github.com/NixOS/rfcs/pull/4

Profpatsch replied:
` is not unicode, though.

nbp commented on lib/types.nix:
```diff
@@ -107,22 +107,80 @@ rec {
       merge = mergeEqualOption;
     };
 
+### INTEGERS
```
nit: `### Integers` , do the same below for other comments. Indent this comment at the same level as the code.

nbp commented on lib/types.nix:
```diff
@@ -107,22 +107,80 @@ rec {
+        merge = mergeOneOption;
+      };
+
+    # specialized subdomains of int
```
nit: `# Specialized ...`

nbp commented on lib/types.nix:
```diff
@@ -196,6 +254,15 @@ rec {
       functor = (defaultFunctor name) // { wrapped = elemType; };
     };
 
+    # TODO: deprecate, like `list'
```
`attrs` is not deprecated! Remove this miss-leading comment.

Profpatsch replied:
In the removed doc commit. I hope I don’t forget to remove it when PR’ing that.

nbp commented on lib/types.nix:
```diff
@@ -107,22 +107,80 @@ rec {
-    # Merge multiple definitions by concatenating them (with the given
-    # separator between the values).
+    /* Merge multiple definitions by concatenating them (with the given
+       separator between the values). */
```
nit: Avoid /* */ comments for small descriptions.

Profpatsch replied:
I’d like to make `/**/` default for “docstrings”, so that they can be distinguished from in-line comments easily.

nbp reviewed (changes requested):
This set of patches moves things around, adding section names, and does *not only* add int types.

Please restrict this PR to adding int types.

Profpatsch committed:
lib/types: fix up documentation formatting a bit

Profpatsch commented:
@nbp I removed the documentation commit that shuffles part of the file around.

nbp reviewed (approved):
nit: There is still a quoting issue in the newly added code.
nit: Changing the quotes of the existing code in unrelated to adding integer types, and as such belong to a different patch.

Profpatsch commented:
Will make an additional PR for the remaining things with wrong quoting style afterwards. It’s mainly in `lib/modules.nix`.
